### PR TITLE
Foundation feature for imaging and cluster creation

### PR DESCRIPTION
**Pull request by bhati-pradeep**
Added data sources :
- nutanix_foundation_discover_nodes
- nutanix_foundation_node_network_details
- nutanix_foundation_nos_packages
- nutanix_foundation_hypervisor_isos
Added resources : 
- nutanix_foundation_image_nodes
- nutanix_foundation_ipmi_config
- nutanix_foundation_image

### CODE REVIEW

**Comment by siddharth-nutanix** [comment]
/ok-to-test
>Acceptance test run status: failure
>

File: client/client.go [hunk 1]
```diff
@@ -45,21 +49,27 @@ type Client struct {
...
+	Username           string
+	Password           string
+	Endpoint           string
+	Port               string
+	Insecure           bool
+	SessionAuth        bool
+	ProxyURL           string
+	FoundationEndpoint string
+	FoundationPort     string
+	RequiredFields     map[string][]string
```
Comment (siddharth-kulshrestha) [inline client/client.go]: Can we also give comments on RequiredFields and Foundation specific fields mentioning why do we require them?

File: client/client.go [hunk 2]
```diff
@@ -69,7 +79,7 @@ type AdditionalFilter struct {
 }
 
 // NewClient returns a new Nutanix API client.
```
Comment (siddharth-kulshrestha) [inline client/client.go]: Please modify this comment `Nutanix API client sounds too vague`

File: client/client.go [hunk 3]
```diff
@@ -132,8 +147,44 @@ func NewClient(credentials *Credentials, userAgent string, absolutePath string)
 	return c, nil
 }
 
+// NewBaseClient returns a basic http/https client based on isHttp flag
+func NewBaseClient(credentials *Credentials, absolutePath string, isHTTP bool) (*Client, error) {
```
Comment (siddharth-kulshrestha) [inline client/client.go:143]: Ideally NewClient should use NewBaseClient as NewClient is doing some operations extra than NewBaseClient.

File: client/client.go [hunk 4]
```diff
@@ -167,12 +218,82 @@ func (c *Client) NewRequest(ctx context.Context, method, urlStr string, body int
...
+	//add api headers
+	req.Header.Add("Content-Type", mediaType)
+	req.Header.Add("Accept", mediaType)
+	req.Header.Add("User-Agent", c.UserAgent)
+
+	return req, nil
+}
+
+// NewUnAuthFormEncodedRequest returns content-type: application/x-www-form-urlencoded based unauth request
+func (c *Client) NewUnAuthFormEncodedRequest(ctx context.Context, method, urlStr string, body map[string]string) (*http.Request, error) {
```
Comment (siddharth-kulshrestha) [inline client/client.go:251]: This looks similar to the previous function. The only difference is that we are using `data.Encode()` we can create another parameter in above function. No need of new function.

File: client/foundation/foundation_file_management_service.go [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+	}
+	listHypervisorISOsResponse := new(ListHypervisorISOsResponse)
+	return listHypervisorISOsResponse, fmo.client.Do(ctx, req, listHypervisorISOsResponse)
+}
+
+//UploadImage uploads the image to foundation vm as per installer type
+func (fmo FileManagementOperations) UploadImage(ctx context.Context, installerType, fileName, source string) (*UploadImageResponse, error) {
+	path := fmt.Sprintf("/upload?installer_type=%s&filename=%s", installerType, fileName)
+
+	// open file. The source should be complete file path eg./Users/...
```
Comment (siddharth-kulshrestha) [inline client/foundation/foundation_file_management_service.go:52]: We should mention this in our documentation that patch should be complete path.

File: client/foundation/foundation_structs.go [hunk 6]
```diff
@@ -0,0 +1,348 @@
...
+	ClusterExternalIP     string   `json:"cluster_external_ip,omitempty"`
+	CvmNtpServers         string   `json:"cvm_ntp_servers,omitempty"`
+	SingleNodeCluster     *bool    `json:"single_node_cluster,omitempty"`
+	ClusterMembers        []string `json:"cluster_members"`
+	CvmDNSServers         string   `json:"cvm_dns_servers,omitempty"`
+	ClusterInitNow        *bool    `json:"cluster_init_now"`
+	HypervisorNtpServers  string   `json:"hypervisor_ntp_servers,omitempty"`
+}
+
+type EosMetadata struct {
```
Comment (siddharth-kulshrestha) [inline client/foundation/foundation_structs.go:88]: Please add comments everywhere.

File: nutanix/config.go [hunk 7]
```diff
@@ -13,27 +14,33 @@ const Version = "3.1"
...
+	Username           string
+	Password           string
+	Port               string
+	Insecure           bool
+	SessionAuth        bool
+	WaitTimeout        int64
+	ProxyURL           string
+	FoundationEndpoint string
+	FoundationPort     string
+	RequiredFields     map[string][]string
```
Comment (siddharth-kulshrestha) [inline nutanix/config.go]: Add comments

File: nutanix/data_source_foundation_hypervisor_isos.go [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+)
+
+func dataSourceFoundationHypervisorIsos() *schema.Resource {
+	return &schema.Resource{
+		ReadContext: dataSourceFoundationHypervisorIsosRead,
+		Schema: map[string]*schema.Schema{
+			"hyperv": {
+				Type:     schema.TypeList,
+				Computed: true,
+				Elem: &schema.Resource{
```
Comment (siddharth-kulshrestha) [inline nutanix/data_source_foundation_hypervisor_isos.go]: Can we move this schema into another function and use everywhere?

File: nutanix/helpers.go [hunk 9]
```diff
@@ -214,6 +216,24 @@ func taskStateRefreshFunc(client *v3.Client, taskUUID string) resource.StateRefr
 	}
 }
 
+func foundationImageRefresh(ctx context.Context, client *foundation.Client, sessionUUID string) resource.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		v, err := client.NodeImaging.ImageNodesProgress(ctx, sessionUUID)
+
+		if err != nil {
+			if strings.Contains(fmt.Sprint(err), "Failed") {
```
Comment (siddharth-kulshrestha) [inline nutanix/helpers.go]: err
```suggestion
			if strings.Contains(err.Error(), "Failed") {
```

File: nutanix/provider.go [hunk 10]
```diff
@@ -1,11 +1,21 @@
...
+	"fmt"
 	"log"
+	"strings"
 
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
 )
 
+var requiredProviderFields map[string][]string = map[string][]string{
+	"prism central": {"username", "password", "endpoint"},
```
Comment (siddharth-kulshrestha) [inline nutanix/provider.go]: any specific reason of using a space in a map key?

File: nutanix/provider.go [hunk 11]
```diff
@@ -136,26 +167,62 @@ func Provider() *schema.Provider {
...
 	}
 }
 
 // This function used to fetch the configuration params given to our provider which
 // we will use to initialize a dummy client that interacts with API.
-func providerConfigure(d *schema.ResourceData) (interface{}, error) {
+func providerConfigure(ctx context.Context, d *schema.ResourceData) (interface{}, diag.Diagnostics) {
 	log.Printf("[DEBUG] config wait_timeout %d", d.Get("wait_timeout").(int))
 
+	disabledProviders := make([]string, 0)
```
Comment (siddharth-kulshrestha) [inline nutanix/provider.go:187]: Why default length is 0?

File: nutanix/provider.go [hunk 12]
```diff
@@ -136,26 +167,62 @@ func Provider() *schema.Provider {
...
+			// for string fields
+			if val, ok := d.Get(attr).(string); ok && val == "" {
+				disabledProviders = append(disabledProviders, k)
+				break
+			}
+			// for integer fields
+			if val, ok := d.Get(attr).(float64); ok && int64(val) == 0 {
+				disabledProviders = append(disabledProviders, k)
+				break
+			}
```
Comment (siddharth-kulshrestha) [inline nutanix/provider.go]: What about boolean and other types of fields? Are we using any integer fields in the above provider auth config?

File: nutanix/provider.go [hunk 13]
```diff
@@ -136,26 +167,62 @@ func Provider() *schema.Provider {
...
+		}
+	}
+
+	if len(disabledProviders) > 0 {
+		diags = append(diags, diag.Diagnostic{
+			Severity: diag.Warning,
+			Summary:  fmt.Sprintf("Disabled Providers: %s. Please provide required fields in provider configuration to enable them. Refer docs.", strings.Join(disabledProviders, ", ")),
```
Comment (siddharth-kulshrestha) [inline nutanix/provider.go:204]: Dont forget to add these details in the docs.

**Review by siddharth-kulshrestha — CHANGES_REQUESTED** [review]

File: client/foundation/foundation_structs.go (hunk 6 above)
Comment (bhati-pradeep) [reply client/foundation/foundation_structs.go:88]: Done

File: client/client.go (hunk 2 above)
Comment (bhati-pradeep) [reply client/client.go]: Done

**Comment by bhati-pradeep** [comment]
Check warnings requirement for prism_central client

File: client/client.go (hunk 3 above)
Comment (bhati-pradeep) [reply client/client.go:143]: Done

File: client/foundation/foundation_file_management_service.go (hunk 5 above)
Comment (bhati-pradeep) [reply client/foundation/foundation_file_management_service.go:52]: Noted

File: nutanix/provider.go (hunk 13 above)
Comment (bhati-pradeep) [reply nutanix/provider.go:204]: Noted

**Comment by bhati-pradeep** [comment]
/ok-to-test
>Acceptance test run status: failure
>

**Review by abhimutant — APPROVED** [review]

**Pull request closed without merging**